### PR TITLE
updated basculechecks so we can log when capabilities don't match up

### PR DESCRIPTION
**Pull request by kristinaspring**
Updated capabilities check so that it can be done either as a check to be enforced, or checked post-enforcer and the results logged.

WIP. need to add:
- [ ] unit tests

### CODE REVIEW

**Comment by johnabass** [comment]
This looks good, but the build is failing

File: basculechecks/basculechecks.go [hunk 1]
```diff
@@ -17,46 +19,177 @@ var (
...
+	if !ok {
+		c.measures.CapabilityCheckOutcome.With(ClientIDLabel, "", PartnerIDLabel, "", OutcomeLabel, RejectedTokenMissing).Add(1)
+		return ErrNoAuth
+	}
+
+	// getting metric information
+	client := token.Principal()
+	partnerIDs, err := getPartnerIDList(token.Attributes())
+	if err != nil {
+		c.measures.CapabilityCheckOutcome.With(ClientIDLabel, client, PartnerIDLabel, "", OutcomeLabel, RejectedUndeterminedPartnerID).Add(1)
```
Comment (joe94) [inline basculechecks/basculechecks.go]: You could add a helper function is you think it'd make things more concise. Something like this?
```
c.addOneForCapabilityOutcome(ClientIDLabel, "", PartnerIDLabel, "", OutcomeLabel, RejectedTokenMissing)
```
```
func (c *capabilityCheck) addOneForCapabilityOutcome(labels...){
    c.measures.CapabilityCheckOutcome.With(labels ...).Add(1)
}
```

File: basculechecks/basculechecks.go [hunk 2]
```diff
@@ -17,46 +19,177 @@ var (
...
+	c := capabilityCheck{
+		prefixToMatch:   matchPrefix,
+		acceptAllMethod: acceptAllMethod,
+		measures:        m,
+	}
+	return &c, nil
+}
 
-		auth, ok := bascule.FromContext(ctx)
+func (c *capabilityCheck) checkCapabilities(capabilities []interface{}, requestInfo bascule.Request) error {
```
[the remark this answers is not in the document]
Comment (kristinaspring) [reply basculechecks/basculechecks.go]: yeah, you can see how much the next version of bascule will help this code!

File: basculechecks/basculechecks.go [hunk 3]
```diff
@@ -17,46 +19,177 @@ var (
...
+	measures        *AuthCapabilityCheckMeasures
+}
+
+var defaultLogger = log.NewNopLogger()
+
+func (c *capabilityCheck) Check(ctx context.Context, token bascule.Token) error {
+	// getting full auth
+	auth, ok := bascule.FromContext(ctx)
+	if !ok {
+		c.measures.CapabilityCheckOutcome.With(ClientIDLabel, "", PartnerIDLabel, "", OutcomeLabel, RejectedTokenMissing).Add(1)
```
[the remark this answers is not in the document]
Comment (kristinaspring) [reply basculechecks/basculechecks.go]: When I asked Jack he said we have to include them, but I'm going to test a few different ways to try and make this code look nicer.

File: basculechecks/basculechecks.go [hunk 4]
```diff
@@ -84,31 +72,28 @@ func (c *capabilityCheck) OnAuthenticated(token bascule.Authentication) {
...
+	if err != nil {
+		counter.With(ReasonLabel, reason).Add(1)
 		return
 	}
-	err = c.checkCapabilities(capabilities, token.Request)
+
+	err = c.checkCapabilities(vals, token.Request)
 	if err != nil {
-		c.measures.CapabilityCheckOutcome.With(ClientIDLabel, client, PartnerIDLabel, partnerID, OutcomeLabel, AcceptedNoCapabilitiesMatch).Add(1)
+		counter.With(ReasonLabel, NoCapabilitiesMatch).Add(1)
```
Comment (joe94) [inline basculechecks/basculechecks.go]: Outcome label should be rejected

File: basculechecks/basculechecks.go [hunk 5]
```diff
@@ -84,31 +72,28 @@ func (c *capabilityCheck) OnAuthenticated(token bascule.Authentication) {
...
 	}
 
-	// getting metric information
-	client := token.Token.Principal()
-	partnerIDs, err := getPartnerIDList(token.Token.Attributes())
+	client, partnerID, reason, err := prepMetrics(token)
+	counter := c.measures.CapabilityCheckOutcome.With(ClientIDLabel, client, PartnerIDLabel, partnerID, OutcomeLabel, AcceptedOutcome)
 	if err != nil {
-		c.measures.CapabilityCheckOutcome.With(ClientIDLabel, client, PartnerIDLabel, "", OutcomeLabel, AcceptedUndeterminedPartnerID).Add(1)
+		counter.With(ReasonLabel, reason).Add(1)
```
Comment (joe94) [inline basculechecks/basculechecks.go]: Outcome label should be rejected

File: basculechecks/metrics.go [hunk 6]
```diff
@@ -2,58 +2,56 @@ package basculechecks
...
-			Type:    xmetrics.HistogramType,
-			Help:    "Difference (in seconds) between time of JWT validation and nbf (including leeway)",
-			Buckets: []float64{-61, -11, -2, -1, 0, 9, 60}, // defines the upper inclusive (<=) bounds
-		},
-		xmetrics.Metric{
-			Name:    EXPHistogram,
-			Type:    xmetrics.HistogramType,
-			Help:    "Difference (in seconds) between time of JWT validation and exp (including leeway)",
-			Buckets: []float64{-61, -11, -2, -1, 0, 9, 60},
+			Help:       "Counter for success and failure reason results through bascule",
```
Comment (joe94) [inline basculechecks/metrics.go]: Update the help message for the capability metrics

**Pull request merged**